src/libexpr/primops.cc

```diff
@@ -1859,12 +1859,12 @@ static void addPath(
         // be rewritten to the actual output).
         state.realiseContext(context);
 
+        StorePathSet refs = StorePathSet();
+
         if (state.store->isInStore(path)) {
             auto [storePath, subPath] = state.store->toStorePath(path);
-            auto info = state.store->queryPathInfo(storePath);
-            if (!info->references.empty())
-                throw EvalError("store path '%s' is not allowed to have references",
-                    state.store->printStorePath(storePath));
+            // FIXME: we should scanForReferences on the path before adding it
```

Comment:
This is just conservatively taking all the references, even if they happened to be filtered out? That seems decent.

Reply:
Yes. This may result in unneeded references, but I think this bug is rare enough that it doesn't matter, at least for now. FIXME is there in case someone stumbles on this in the future and wants to fix it :)

Reply:
It's not ideal since it allows untrusted users to add paths with incorrect references to the store. But then that's already the case with `addTextToStore()`...

Reply:
> It's not ideal since it allows untrusted users to add paths with incorrect references to the store

Hm, I'm not sure I see how this is possible? The references here must come from a path that's already in store, and therefore has been `scanForReferences` once already, it's just that the user may filter it so some references are not actually present in the resulting path while being part of path info as far as nix is concerned.

Reply:
The issue isn't whether the references are correct here, but that `RemoteStore::addToStoreFromDump` trusts whatever the client passes in. So you could write an untrusted client that adds paths with incorrect references.